File: lib/reentrySignals.ts
import { ndotIndicatesDecay } from './satelliteHelpers';
import type { ObjectTrend, TleEntry } from './types';

export function isDebrisEntry(entry: TleEntry): boolean {
  const nameUpper = entry.name.toUpperCase();
  return (
    entry.isDebris ||
    nameUpper.includes('DEB') ||
    nameUpper.includes('DEBRIS')
  );
}

export function decayAltKmFromTrend(trend: ObjectTrend): number {
  if (trend.smaLatest) return Math.max(0, trend.smaLatest - 6378.137);
  return trend.perigeeLatest ?? 0;
}

type SignalInput = {
  bstarSlope14d: number | null;
  ndotSlope14d: number | null;
  ndotLatest: number | null;
  ndotMean14d: number | null;
  perigeeSlope14d: number | null;
  smaSlope14d: number | null;
  decayAltKm: number;
};

export function decaySignalFlags(input: SignalInput) {
  const bstarAgrees = (input.bstarSlope14d ?? 0) > 0;
  const ndotAgrees =
    (input.ndotSlope14d !== null && input.ndotSlope14d > 0) ||
    (input.ndotLatest !== null &&
      ndotIndicatesDecay(input.ndotLatest, input.decayAltKm)) ||
    (input.ndotMean14d !== null &&
      ndotIndicatesDecay(input.ndotMean14d, input.decayAltKm));
  const altAgrees =
    (input.perigeeSlope14d ?? 0) < -0.01 ||
    (input.smaSlope14d ?? 0) < -0.01;

  return { bstarAgrees, ndotAgrees, altAgrees };
}

export function allTrendSignalsAgree(trend: ObjectTrend): boolean {
  const decayAltKm = decayAltKmFromTrend(trend);
  const flags = decaySignalFlags({
    bstarSlope14d: trend.bstarSlope14d,
    ndotSlope14d: null,
    ndotLatest: trend.meanMotionDotLatest,
    ndotMean14d: trend.meanMotionDotMean14d,
    perigeeSlope14d: trend.perigeeSlope14d,
    smaSlope14d: trend.smaSlope14d,
    decayAltKm,
  });
  return flags.bstarAgrees && flags.ndotAgrees && flags.altAgrees;
}

export function trendSignalsAgree(trend: ObjectTrend): boolean {
  const decayAltKm = decayAltKmFromTrend(trend);
  const flags = decaySignalFlags({
    bstarSlope14d: trend.bstarSlope14d,
    ndotSlope14d: null,
    ndotLatest: trend.meanMotionDotLatest,
    ndotMean14d: trend.meanMotionDotMean14d,
    perigeeSlope14d: trend.perigeeSlope14d,
    smaSlope14d: trend.smaSlope14d,
    decayAltKm,
  });
  return [flags.bstarAgrees, flags.ndotAgrees, flags.altAgrees].filter(
    Boolean
  ).length >= 2;
}

export function allSignalsAgreeFromSlopes(
  input: SignalInput
): boolean {
  const flags = decaySignalFlags(input);
  return flags.bstarAgrees && flags.ndotAgrees && flags.altAgrees;
}

File: backend/compute/reentry_signals.py
"""Ported from lib/reentrySignals.ts (plan §17 Phase 3 + Phase 4).

Phase 3 needed only decay_signal_flags/all_signals_agree_from_slopes (the
slope-only helpers explainReentryTrend calls). Phase 4 needs the rest:
is_debris_entry, decay_alt_km_from_trend, all_trend_signals_agree, and
trend_signals_agree -- the ObjectTrend-facing wrappers resolveReentryRisk()
and objectTrendToReentryRisk() call directly.
"""
from __future__ import annotations

from compute.satellite_helpers import ndot_indicates_decay

EARTH_RADIUS_KM = 6378.137


def is_debris_entry(entry: dict) -> bool:
    name_upper = entry["name"].upper()
    return bool(entry.get("isDebris")) or "DEB" in name_upper or "DEBRIS" in name_upper


def decay_alt_km_from_trend(trend: dict) -> float:
    sma_latest = trend.get("smaLatest")
    if sma_latest:
        return max(0.0, sma_latest - EARTH_RADIUS_KM)
    return trend.get("perigeeLatest") or 0


def decay_signal_flags(input: dict) -> dict:
    bstar_slope_14d = input.get("bstarSlope14d")
    ndot_slope_14d = input.get("ndotSlope14d")
    ndot_latest = input.get("ndotLatest")
    ndot_mean_14d = input.get("ndotMean14d")
    perigee_slope_14d = input.get("perigeeSlope14d")
    sma_slope_14d = input.get("smaSlope14d")
    decay_alt_km = input["decayAltKm"]

    bstar_agrees = (bstar_slope_14d if bstar_slope_14d is not None else 0) > 0
    ndot_agrees = (
        (ndot_slope_14d is not None and ndot_slope_14d > 0)
        or (
            ndot_latest is not None
            and ndot_indicates_decay(ndot_latest, decay_alt_km)
        )
        or (
            ndot_mean_14d is not None
            and ndot_indicates_decay(ndot_mean_14d, decay_alt_km)
        )
    )
    alt_agrees = (
        (perigee_slope_14d if perigee_slope_14d is not None else 0) < -0.01
        or (sma_slope_14d if sma_slope_14d is not None else 0) < -0.01
    )

    return {
        "bstarAgrees": bstar_agrees,
        "ndotAgrees": ndot_agrees,
        "altAgrees": alt_agrees,
    }


def _trend_signal_input(trend: dict, decay_alt_km: float) -> dict:
    return {
        "bstarSlope14d": trend.get("bstarSlope14d"),
        "ndotSlope14d": None,
        "ndotLatest": trend.get("meanMotionDotLatest"),
        "ndotMean14d": trend.get("meanMotionDotMean14d"),
        "perigeeSlope14d": trend.get("perigeeSlope14d"),
        "smaSlope14d": trend.get("smaSlope14d"),
        "decayAltKm": decay_alt_km,
    }


def all_trend_signals_agree(trend: dict) -> bool:
    decay_alt_km = decay_alt_km_from_trend(trend)
    flags = decay_signal_flags(_trend_signal_input(trend, decay_alt_km))
    return flags["bstarAgrees"] and flags["ndotAgrees"] and flags["altAgrees"]


def trend_signals_agree(trend: dict) -> bool:
    decay_alt_km = decay_alt_km_from_trend(trend)
    flags = decay_signal_flags(_trend_signal_input(trend, decay_alt_km))
    return sum(1 for v in flags.values() if v) >= 2


def all_signals_agree_from_slopes(input: dict) -> bool:
    flags = decay_signal_flags(input)
    return flags["bstarAgrees"] and flags["ndotAgrees"] and flags["altAgrees"]
